tensorboard/components/vz_line_chart2/log-scale.ts

```diff
@@ -0,0 +1,164 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+namespace vz_line_chart2 {
+
+const MIN_VALUE = 1e-15;
+
+function log(x: number): number {
+  return Math.log(x) / Math.log(10);
+}
+
+function pow(x: number): number {
+  return Math.pow(10, x);
+}
+
+export class LogScale extends TfScale {
+  private _d3LogScale = d3.scaleLog();
+  private _untransformedDomain: number[];
+
+  constructor() {
+    super();
+    this.padProportion(.2);
+  }
+
+  public scale(x: number): number {
+    // Returning NaN makes sure line plot does not plot illegal values.
+    if (x <= 0) return NaN;
+    return this._d3LogScale(x);
+  }
+
+  public invert(x: number): number {
+    return this._d3LogScale.invert(x);
+  }
+
+  public scaleTransformation(value: number) {
+    return this.scale(value);
+  }
+
+  public invertedTransformation(value: number) {
+    return this.invert(value);
+  }
+
+  public getTransformationDomain(): [number, number] {
+    return this.domain() as [number, number];
+  }
+
+  protected _getDomain() {
+    return this._untransformedDomain;
+  }
+
+  protected _setDomain(values: number[]) {
+    this._untransformedDomain = values;
+    const [min, max] = values;
+    super._setDomain([Math.max(MIN_VALUE, min), max]);
+  }
+
+  /**
+   * Given a domain, pad it and clip the lower bound to MIN_VALUE.
```

Comment:
>  It seems like if at all possible we should avoid introducing thresholds like this that distort the mathematical soundness, since that's the whole motivation for replacing ModifiedLogScale in the first place.

JavaScript does not have infinite precision and starts to behave very oddly starting from one point. 
```
> 1e-16 * 1e-30
< 1e-46
> 1e-16 * 1e-40
< 9.999999999999999e-57
> 1e-16 * 1e-50
< 1e-66
> 1e-15 * 1e-1
< 1.0000000000000001e-16
```

I decided to put some arbitrary precision without reading too much about the ECMAScript spec but regardless, it seems like there is a lower bound to `Math.log10(x)` before it is regarded as negative infinity. I will make sure the MIN_VALUE is smaller but it is inevitable that we have some arbitrary value where we clip.

Reply:
The odd behavior there looks like regular floating point representation errors - the actual values are still essentially correct up to minute precision.  Fair enough that some clipping is required but I do think it's better to have it be as small as technically possible vs. just 10^-15.